book/src/rfcs/foreign_field_mul.md

```diff
@@ -0,0 +1,1273 @@
+- Have copy constraints for $q'_{01}$ and $q'_2$ as outputs of the `ForeignFieldMul` gate and inputs to the `multi-range-check` gadget
+
+Note that since the foreign field multiplication gate computes $q'$ from $q$ which is already in the witness and $q'_{01}$ and $q'_2$ have copy constraints to a `multi-range-check` gadget that fully constrains their decomposition from $q'$, then the `ForeignFieldMul` gate does not need to store an additional copy of $q'_0$ and $q'_1$.
+
+### An optimization
+
+Since the $q < f$ and $r < f$ checks must be done for each multiplication it makes sense to integrate them into the foreign field multiplication gate. By doing this we can save 4 rows per multiplication.
+
+Doing this doesn't require adding a lot more witness data because the operands for the bound computations $q' = q + f'$ and $r' = r + f'$ are already present in the witness of the multiplication gate.  We only need to store the bounds $q'$ and $r'$ in permutable witness cells so that they may be copied to multi-range-check gates to check they are each less than $2^t$.
+
+To constrain $x + f' = x'$, the equation we use is
+
+$$
+x + 2^t = \mathcal{o} \cdot f + x',
+$$
+
+where $x$ is the original value, $\mathcal{o}=1$ is the field overflow bit and $x'$ is the remainder and our desired addition result (e.g. the bound).  Rearranging things we get
+
+$$
+x + 2^t - f = x',
+$$
+
+which is just
+
+$$
+x + f' = x',
+$$
+
+Recall from the section "Avoiding borrows" that $f'$ is often larger than $f$.  At first this seems like it could be a problem because in multiplication each operation must be less than $f$.  However, this is because the maximum size of the multiplication was quadratic in the size of $f$ (we use the CRT, which requires the bound that $a \cdot b < 2^t \cdot n$).  However, for addition the result is much smaller and we do not require the CRT nor the assumption that the operands are smaller than $f$.  Thus, we have plenty of space in $\ell$-bit limbs to perform our addition.
+
+So, the equation we need to constrain is
+
+$$
+x + f' = x'.
+$$
+
+We can expand the left hand side into the 2 limb format in order to obtain 2 intermediate sums
+
+$$
+\begin{aligned}
+s_{01} = x_{01} + f_{01}' \\
+s_2 = x_2 + f'_2 \\
+\end{aligned}
+$$
+
+where $x_{01}$ and $f'_{01}$ are defined like this
+
+$$
+\begin{aligned}
+x_{01} = x_0 + 2^{\ell} \cdot x_1 \\
+f'_{01} = f'_0 + 2^{\ell} \cdot f'_1 \\
+\end{aligned}
+$$
+
+and $x$ and $f'$ are defined like this
+
+$$
+\begin{aligned}
+x = x_{01} + 2^{2\ell} \cdot x_2 \\
+f' = f'_{01} + 2^{2\ell} \cdot f'_2 \\
+\end{aligned}
+$$
+
+Going back to our intermediate sums, the maximum bit length of sum $s_{01}$ is computed from the maximum bit lengths of $x_{01}$ and $f'_{01}$
+
+$$
+\underbrace{(2^{\ell} - 1) + 2^{\ell} \cdot (2^{\ell} - 1)}_{x_{01}} + \underbrace{(2^{\ell} - 1) + 2^{\ell} \cdot (2^{\ell} - 1)}_{f'_{01}} = 2^{2\ell+ 1} - 2,
```

Comment:
the subindices do not seem to be properly parsed, but perhaps they will when included in the mdbook

Reply:
Yeah, it looks fine in mdbook, but sadly github latex has some bugs.